frontend/controller/serviceworkers/sw-primary.js

```diff
@@ -1,12 +1,148 @@
+} from '../../utils/events.js'
+import './push.js'
+
+deserializer.register(GIMessage)
+deserializer.register(Secret)
+
 // https://serviceworke.rs/message-relay_service-worker_doc.html
 // https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API/Using_Service_Workers
 // https://jakearchibald.com/2014/using-serviceworker-today/
 // https://github.com/w3c/ServiceWorker/blob/master/explainer.md
 // https://frontendian.co/service-workers
 // https://stackoverflow.com/a/49748437 => https://medium.com/@nekrtemplar/self-destroying-serviceworker-73d62921d717 => https://love2dev.com/blog/how-to-uninstall-a-service-worker/
 
+const reducer = (o, v) => { o[v] = true; return o }
+// Domains for which debug logging won't be enabled.
+const domainBlacklist = [
+  'sbp',
+  'okTurtles.data'
+].reduce(reducer, {})
+// Selectors for which debug logging won't be enabled.
+const selectorBlacklist = [
+  'chelonia/db/get',
+  'chelonia/db/set',
+  'chelonia/rootState',
+  'chelonia/haveSecretKey',
+  'chelonia/private/enqueuePostSyncOps',
+  'chelonia/private/invoke',
+  'state/vuex/state',
+  'state/vuex/getters',
+  'state/vuex/settings',
+  'gi.db/settings/save',
+  'gi.db/logs/save'
+].reduce(reducer, {})
+sbp('sbp/filters/global/add', (domain, selector, data) => {
+  if (domainBlacklist[domain] || selectorBlacklist[selector]) return
+  console.debug(`[sbp] ${selector}`, data)
+})
+
+// These are all of the events that will be forwarded to all open tabs and windows
+;[
+  CHELONIA_RESET, CONTRACTS_MODIFIED, CONTRACT_IS_SYNCING, EVENT_HANDLED, LOGIN,
+  LOGIN_ERROR, LOGOUT, ACCEPTED_GROUP, CHATROOM_USER_STOP_TYPING,
+  CHATROOM_USER_TYPING, DELETED_CHATROOM, LEFT_CHATROOM, LEFT_GROUP,
+  JOINED_CHATROOM, JOINED_GROUP, KV_EVENT, MESSAGE_RECEIVE, MESSAGE_SEND,
+  NAMESPACE_REGISTRATION, NEW_CHATROOM_UNREAD_POSITION, NEW_LAST_LOGGED_IN,
+  NEW_PREFERENCES, NEW_UNREAD_MESSAGES, NOTIFICATION_EMITTED,
+  NOTIFICATION_REMOVED, NOTIFICATION_STATUS_LOADED, OFFLINE, ONLINE,
+  PROPOSAL_ARCHIVED, SERIOUS_ERROR, SWITCH_GROUP
+].forEach(et => {
+  sbp('okTurtles.events/on', et, (...args) => {
+    const { data } = serializer(args)
+    const message = {
+      type: 'event',
+      subtype: et,
+      data
+    }
+    self.clients.matchAll()
+      .then((clientList) => {
+        clientList.forEach((client) => {
+          client.postMessage(message)
+        })
+      })
+  })
+})
+
+// Logs are treated especially to avoid spamming logs with event emitted
+// entries
+sbp('okTurtles.events/on', CAPTURED_LOGS, (...args) => {
+  const { data } = serializer(args)
+  const message = {
+    type: CAPTURED_LOGS,
+    data
+  }
+  self.clients.matchAll()
+    .then((clientList) => {
+      clientList.forEach((client) => {
+        client.postMessage(message)
+      })
+    })
+})
+
+sbp('sbp/selectors/register', {
+  'state/vuex/state': () => sbp('chelonia/rootState'),
+  'state/vuex/getters': () => {
+    const obj = Object.create(null)
```

Comment:
Instead of creating this object each time `'state/vuex/getters'` is called, does it make sense to create a singleton?

Reply:
> does it make sense to create a singleton?

It does